cdc/scheduler/agent.go

```diff
@@ -0,0 +1,448 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scheduler
+
+import (
+	"sync"
+	"time"
+
+	"github.com/edwingeng/deque"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/ticdc/cdc/model"
+	"github.com/pingcap/ticdc/cdc/scheduler/util"
+	"github.com/pingcap/ticdc/pkg/context"
+	cerrors "github.com/pingcap/ticdc/pkg/errors"
+	"github.com/uber-go/atomic"
+	"go.uber.org/zap"
+)
+
+// Agent is an interface for an object inside Processor that is responsible
+// for receiving commands from the Owner.
+// Ideally the processor should drive the Agent by Tick.
+type Agent interface {
+	// Tick is called periodically by the processor to drive the Agent's internal logic.
+	Tick(ctx context.Context) error
+
+	// GetLastSentCheckpointTs returns the last checkpoint-ts already sent to the Owner.
+	GetLastSentCheckpointTs() (checkpointTs model.Ts)
+}
+
+// TableExecutor is an abstraction for "Processor".
+type TableExecutor interface {
+	AddTable(ctx context.Context, tableID model.TableID) error
+	RemoveTable(ctx context.Context, tableID model.TableID) (done bool, err error)
+	IsAddTableFinished(ctx context.Context, tableID model.TableID) (done bool)
+	IsRemoveTableFinished(ctx context.Context, tableID model.TableID) (done bool)
+
+	GetAllCurrentTables() []model.TableID
```

Comment:
Better to add comments to this function, seems it returns running, adding, removing tables.